aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp

```diff
@@ -1958,78 +1958,62 @@ static void lu_factor_batched_magma(const Tensor& input, const Tensor& pivots, c
     } else {
-      // There is a bug in lu_factor_batched_magma in MAGMA < 2.5.2, see
-      // https://bitbucket.org/icl/magma/issues/13/getrf_batched-kernel-produces-nans-on
-      std::tuple<magma_int_t, magma_int_t, magma_int_t> version;
-      magma_version(&std::get<0>(version), &std::get<1>(version), &std::get<2>(version));
-      if (version >= std::make_tuple<magma_int_t, magma_int_t, magma_int_t>(2, 5, 2)) {
-        lu_factor_batched_magma(input, pivots, infos, compute_pivots);
-      } else {
-        lu_factor_looped_magma(input, pivots, infos, compute_pivots);
-      }
+      lu_factor_batched_magma(input, pivots, infos, compute_pivots);
     }
   };
-#endif
 
+  const auto preferred_backend = at::globalContext().linalgPreferredBackend();
+#ifdef USE_CUSOLVER
+  const auto lu_factor_cusolver = [batch_size, m, n](const Tensor& input, const Tensor& pivots, const Tensor& infos, bool compute_pivots) {
+    if (batch_size == 1 || m != n || m >= 512) {
+      lu_factor_looped_cusolver(input, pivots, infos, compute_pivots);
+    } else {
+      lu_factor_batched_cublas(input, pivots, infos, compute_pivots);
+    }
+  };
+
+  if (preferred_backend == at::LinalgBackend::Cusolver) {
+    lu_factor_cusolver(input, pivots, infos, compute_pivots);
+  } else
+#endif // ifdef USE_CUSOLVER
+  if (preferred_backend == at::LinalgBackend::Magma) {
+    lu_factor_magma(input, pivots, infos, compute_pivots);
+  } else {  // preferred backend == default
 #ifdef USE_CUSOLVER
-  auto preferred_backend = at::globalContext().linalgPreferredBackend();
-  switch (preferred_backend) {
-    case at::LinalgBackend::Cusolver:
-      lu_factor_looped_cusolver(input, pivots, infos, compute_pivots, use_magma_);
-      break;
-    case at::LinalgBackend::Magma:
-#if AT_MAGMA_ENABLED()
-      lu_factor_magma(input, pivots, infos, compute_pivots);
-      break;
-#endif
-    default:
 #if AT_MAGMA_ENABLED()
-      // We do not use cuSOLVER for complex inputs if !get_pivots since nan_to_num_ does not work with it.
-      // See https://github.com/pytorch/pytorch/issues/59247 for more info
-      // Provided the above, use a heuristic to determine that cusolver is faster than MAGMA
-      const auto m = input.size(-2);
-      const auto use_cusolver = ((batch_size == 1 || (batch_size <= 8 && m <= 16))
-                                 && (!input.is_complex() || compute_pivots));
-      if (use_cusolver) {
-        lu_factor_looped_cusolver(input, pivots, infos, compute_pivots, use_magma_);
-      } else {
-        lu_factor_magma(input, pivots, infos, compute_pivots);
-      }
+    // If magma batched is buggy, we use cusolver
+    // otherwise, lu_factor just works for square matrices, for non-square matrices magma batched is the fastest
+    // otherwise (i.e. for square matrices), we choose between cusolver and magma using a heuristic
+    if (magma_batched_buggy || (m == n && (m <= 16 || (m <= 128 && batch_size <= 16)))) {
```

Comment:
In the benchmark results from the issue description for size `(1, 2048, 2048)`
magma batched: 29860 μs
cusolver: 12920 μs
cusolver is 2.3 times faster, but it won't be chosen with current conditions.

Reply:
The thing is that:
- magma is buggy for a number of inputs
- cusolver does not work for non-square inputs

Even more, this heuristic, although simpler, is worse than the current one.
I can add: "choose cublas/cusolver if m == 1" as well, and that would account for this case.

Reply:
> magma is buggy for a number of inputs

Which ones?

> cusolver does not work for non-square inputs

cusolver works for non-square inputs. You confused it with cublas which doesn't work with non-squares.

Reply:
yes, sorry, cublas, the faster of the two. For magma, see the linked bitbucket issue in the comment.

Reply:
For reference:   // https://bitbucket.org/icl/magma/issues/13/getrf_batched-kernel-produces-nans-on

Reply:
tbf, this was fixed almost 3 years ago. I agree that's fine to do as if this error never existed if we don't use those libraries when building PyTorch or in CI.

Reply:
Windows CI uses 2.5.4 and Linux CI uses 2.6.1